docs/_data/nakadi-event-bus-api.yaml

```diff
@@ -2469,6 +2489,40 @@ definitions:
 
           If this is set it MUST be a valid required field as defined in the schema.
 
+      cleanup_policy:
+        type: string
+        x-extensible-enum:
+          - delete
+          - compact
+        default: 'delete'
+        description: |
+          Event type cleanup policy. There are two possible values:
+
+          - 'delete' cleanup policy will delete old events after retention time expires. Nakadi guarantees that each
+            event will be available for at least the retention time period. However Nakadi doesn't guarantee that event
+            will be deleted right after retention time expires.
+
+          - 'compact' cleanup policy will keep only the latest event for each event key. The compaction is performed per
```

Comment:
I feel it would be better, and less burden on producing clients, if the `partition_compaction_key` was configured as part of the `EventType` (similar to how `partition_key_fields` is done today) - this then would mean existing producing applications would need have zero change to take advantage of this capability (as it would only need to be a change in the specification of the Event Type itself), as well as being a more symmetric API (having both partitioning and compaction key data defined/handled in the same way would be nice).

Another big benefit of having the `compaction_key` defined in the event-type is that existing Nakadi clients will continue to work for event production - adding to the event metadata for every event may require client library upgrades to allow client code upgrades.

Reply:
@conorclifford our first version was exactly how you describe and exactly for the same reasons. We later on changed to this new approach. I'll try to explain why and then we can maybe agree on the change or we can revert it back.

There are some problems with the approach used for `partition_key_fields`:

1. non-injective functions: users might specify complete sections an event as values for hashing, or in this case compaction: Arrays, entire objects and numbers. In the case of partitioning, having a function that is non-injective is not an issue for partitioning, since it would only mean that 2 events would be stored in the same partition. As for compaction, the function needs to be strictly injective. The consequences for having different set of values mapping to the same compaction key would be undesirable deletion of events with conflicting keys. It's very hard to implement a function that is injective given the domain is so complex. I can tell for example that numbers are problematic due to precision. In Java, if a user decides to compact two events `{"foo": 0.99999999999999999}` and `{"foo":1}` in some cases, both number could be mapped to the same value, duo to internal limitations of how Java represents values internally. I know for sure that our implementation transforms `0.0` into `0` and that would definitely cause a conflict, which might be accidental. Only strings are considered to be 100% safe for implementing an injective function. That would lead to an asymmetry, since partition key fields and compaction fields would have different constraints in the types of values chosen.

2. Difficulty to debug: since the computed compaction key would only exist in Kafka, it would be hard to explain to users why two events are compacted, since Nakadi API would not expose the generated value. One could suggest to enrich the event metadata with the generated value, so that it becomes easier to debug. That's also a possibility. But then we would have to make it clear if this fields is always controlled by Nakadi or if users can also set it themselves.

I would say point 2 (debugging) is less of a problem, since we already know about one solution. Problem number 1 (non-injective function) is the most complex in my opinion, since it could lead to data loss. There is no solution to the number precision problem since this is a Json limitation regarding precise definition of what a number can be. If we document it, someone will not be aware of it and will hit the problem. Limiting it to strings could be a solution. But we considered that this feature should be taken more seriously then partitioning and that it would be reasonable to expect producer to change their code in order to generate this key in a well though way using only strings.

Please, tell us what do you think about these two problems and how we approached it.

Reply:
@conorclifford I forgot to mention that this key should have clearly defined maximum size. When the key is a result of an internal Nakadi computation, it could be that users would have problems on runtime in random ways, depending on the combination of values used to generate the key.

Reply:
I was unaware of this complex semantics with `partition_key_fields` (out of interest, do you know what proportion of event-types are taking advantage of this capability with complex objects, vs. those simply providing a string field?)

With that said, I still think having the "compaction" key defined within the event-type would give more consistency with partitioning - enforcing that this key be a "string" as part of the event-type JSON Schema validation would meet the requirements for the value. 

Even taking it one step further, and mandating that for log compacted topics that the `partition_key_field` be a string, and that that be the value that is used for compaction, could simply remove any possible confusing situations where hashed partition assignment and compaction keys do not align as expected by clients (it would, however, lose the ability to have different compaction vs. partitioning, which can be beneficial in certain (somewhat edge) cases).

Regarding the `defined maximum size` - I'm not sure what you mean here - defining the compaction key as a string (whether it is in the event-type, or as in this proposal, in the metadata) there wouldn't be any "internal nakadi computation", would there? (just the string itself being used as the key, no?)

Reply:
@conorclifford I agree with that narrowing the possible values to strings would solve the problem 👍 I'll catch up with the team and see if we can implement this way.

Regarding maximum size there is a small computation on our side. We would collect all the strings mentioned in the partition_compaction_keys and generate a json array and then serialize it as a string. That would be the key used at Kafka level. But I just checked yesterday and Kafka has no limit regarding the size of this key. So this problem does not exist =)

Reply:
@conorclifford we had a small chat internally and even though we all agree that it would be feasible to implement an injective function by limiting the value to string, we see other problems with this approach.

We would have to check that the fields are present in the schema (which we can do already, but not without problems. Some paths are not accessible, like "anyOf", for example). We would have to extend this check to verify that these fields are marked as "required" which is not in place today. We would also have to check that all the values are not empty strings.

If the publisher then sends some empty string, we would have to detect that and properly inform them that it was not possible to publish the event because one of values they have specified in the path was empty. Which is a problem harder to clearly communicate than just saying that  `partition_compaction_key` is empty.

In the future, we would be required to implement deletion of events from log compacted event types. If we have a single string as a key for compaction, the api for deletion would be much simpler than if we would require from the user a set of fields matching a given structure.

We would like to know if changing the client library is the major pain point for your team or if the main problem you see with the proposed approach is the asymmetry in the with partitioning.

Reply:
@rcillo interesting points, especially the deletion one.

In our group alone there's a few different applications that will need enhancing to support this, some are directly interacting with the HTTP API (so direct code changes for this at least), whereas others depend on one or more of the existing client libs (its quite likely that "nakadi-java" is the most used lib) - its by no means a blocker, just a bit of additional work to be done.

Reply:
@conorclifford yes, we actually even started to consider that the partitioning keys feature might have problems so we would like not to replicate that approach again. It's an interesting situation where designing a feature we learned something about existing feature that is not ideal.

So, we truly believe that this new approach will be rock solid and work. We would like to proceed and deploy a version for you to test =)

Regarding the additional work, since our goal is to help your team achieve your goals, we can even consider a scenario where we contribute to nakadi-java. If that could speed things up for your team, of course. We believe that the additional work will pay off in terms of stability and less bug fixes or incidents.

Reply:
@rcillo  for now, there are some topics we can start with that only require changes to a single in-team repo, so the nakadi-java client change should not block any early testing.